ecommerce/tests/test_views.py

```diff
@@ -1,44 +1,305 @@
+        response = self.client.get('/ecommerce/search')
+        self.assertEqual(response.status_code, 200)
+           
+    def test_view_url_accessible_by_name(self):
+        response = self.client.get(reverse('product_search'))
+        self.assertEqual(response.status_code, 200)
+
+    def test_uses_correct_template(self):
+        response = self.client.get(reverse('product_search'))
+        self.assertEqual(response.status_code, 200)
+
+        # Check we used correct template
+        self.assertTemplateUsed(response, 'ecommerce/product_list.html')
+
+class ProductGetViewTest(TestCase):
+    def test_view_url_exists_at_desired_location(self):
+        response = self.client.get('/ecommerce/')
+        self.assertEqual(response.status_code, 200)
+           
+    def test_view_url_accessible_by_name(self):
+        response = self.client.get(reverse('products'))
+        self.assertEqual(response.status_code, 200)
+
+    def test_uses_correct_template(self):
+        response = self.client.get(reverse('products'))
+        self.assertEqual(response.status_code, 200)
+
+        # Check we used correct template
+        self.assertTemplateUsed(response, 'ecommerce/product_list.html')
+
+class CartCheckoutViewTest(TestCase):
+    def setUp(self):
+        test_user = User.objects.create_user(username='testuser', password='1X<ISRUkw+tuK')
+        test_user.save()
+
+        self.test_product = ProductFactory(quantity='0')
+        self.test_cart = CartFactory(product=self.test_product, quantity='1')
+           
+    def test_view_url_accessible_by_name(self):
+        response = self.client.get(reverse('cart_checkout'))
+        self.assertEqual(response.status_code, 302)
+
+    def test_redirect_if_not_logged_in(self):
+        response = self.client.get(reverse('cart_checkout'))
+        self.assertEqual(response.status_code, 302)
+        self.assertTrue(response.url.startswith('/accounts/login/'))
+
+    def test_logged_in(self):
+        login = self.client.login(username = 'testuser', password = '1X<ISRUkw+tuK')
+        response = self.client.get(reverse('cart_checkout'))
+        self.assertEqual(response.status_code, 200)
+
+    def test_uses_correct_template(self):
+        self.client.login(username = 'testuser', password = '1X<ISRUkw+tuK')
+        response = self.client.get(reverse('cart_checkout'))
+        self.assertEqual(response.status_code, 200)
+
+        # Check we used correct template
+        self.assertTemplateUsed(response, 'ecommerce/checkout.html')
+
+    def test_product_quantity_validate(self):
+        self.client.login(username = 'testuser', password = '1X<ISRUkw+tuK')
+        self.client.get(reverse('cart_add', kwargs={'pk': self.test_product.pk}))
+        response = self.client.get(reverse('cart_checkout'))
+        print(response)
+        self.assertEqual(response.status_code, 200)
+        # self.assertRedirects(response, 'ecommerce/checkout.html')
```

Comment:
self.test_product = ProductFactory(quantity='0')
self.test_cart = CartFactory(product=self.test_product, quantity='1')
=> như này thì sẽ rơi vào trường hợp xử lý lỗi elif item_quantity > product_stock
```suggestion
        self.assertRedirects(response, '/ecommerce/cart', status_code=200, msg_prefix="...")
```